rust/src/builder.rs

```diff
@@ -57,21 +144,34 @@ where
 
     /// Returns the current capacity of the builder (number of elements)
     pub fn capacity(&self) -> i64 {
-        let byte_capacity = self.buffer.capacity();
-        (byte_capacity / mem::size_of::<T>()) as i64
+        let byte_capacity = self.buffer.capacity() as i64;
+        byte_capacity * 8
     }
 
     /// Push a value into the builder, growing the internal buffer as needed
-    pub fn push(&mut self, v: T) -> Result<()> {
+    pub fn push(&mut self, v: bool) -> Result<()> {
         self.reserve(1)?;
-        self.write_bytes(v.to_byte_slice(), 1)
+        if v {
+            // Check needed here to increase the buffers `len` to ensure that `set_bit` is safe
+            // Zeroing out other bits also makes testing easier
+            if self.len > self.buffer.len() as i64 * 8 {
+                // safe to `unwrap` here as `write` performs a `capacity` check which is
+                // already done by `reserve`
+                self.buffer.write(&[0]).unwrap();
+            }
+            bit_util::set_bit(self.buffer.data_mut(), (self.len) as usize);
+        }
+        self.len += 1;
+        Ok(())
     }
 
     /// Push a slice of type T, growing the internal buffer as needed
-    pub fn push_slice(&mut self, slice: &[T]) -> Result<()> {
-        let array_slots = slice.len() as i64;
-        self.reserve(array_slots)?;
-        self.write_bytes(slice.to_byte_slice(), array_slots)
+    pub fn push_slice(&mut self, slice: &[bool]) -> Result<()> {
+        let array_slots = slice.len();
+        for i in 0..array_slots {
```

Comment:
This is fine but it may be too slow if we write bit by bit. We can optimize this in future though.